velox/functions/prestosql/types/BigintEnumType.cpp

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <folly/Synchronized.h>
+
+#include "velox/functions/prestosql/types/BigintEnumType.h"
+
+namespace facebook::velox {
+
+namespace {
+std::unordered_map<int64_t, std::string> toFlippedMap(
+    std::unordered_map<std::string, int64_t> map,
+    std::string name) {
+  std::unordered_map<int64_t, std::string> flippedMap;
+  for (const auto& [key, value] : map) {
+    bool ok = flippedMap.emplace(value, key).second;
+    VELOX_USER_CHECK(
+        ok, "Invalid enum type {}, contains duplicate value {}", name, value);
+  }
+  return flippedMap;
+}
+
+std::string orderedMapToString(
+    const std::map<std::string, int64_t>& sortedMap) {
+  // The values map should be sorted when converting to string for 2 reasons:
+  // 1. Printing the type will be consistent.
+  // 2. The enum name + values map is used as the key in the cache for get(), so
+  // the string should be consistent.
+  std::ostringstream oss;
+  oss << "{";
+  for (auto it = sortedMap.begin(); it != sortedMap.end(); ++it) {
+    if (it != sortedMap.begin()) {
+      oss << ", ";
+    }
+    oss << "\"" << it->first << "\"" << ": " << it->second;
+  }
+  oss << "}";
+  return oss.str();
+}
+
+std::string mapToString(
+    const std::unordered_map<std::string, int64_t>& unorderedMap) {
+  std::map<std::string, int64_t> sortedMap(
+      unorderedMap.begin(), unorderedMap.end());
+  return orderedMapToString(sortedMap);
+}
+} // namespace
+
+// Should only be called from get() to create a new instance.
+BigintEnumType::BigintEnumType(const LongEnumParameter& parameters)
+    : parameters_{TypeParameter(parameters)},
+      name_{parameters.name},
+      flippedMap_{toFlippedMap(parameters.valuesMap, name_)} {}
+
+std::string BigintEnumType::toString() const {
+  std::map<std::string, int64_t> sortedMap;
+  for (const auto& [key, value] : flippedMap_) {
+    sortedMap[value] = key;
+  }
+  return fmt::format("{}:BigintEnum({})", name_, orderedMapToString(sortedMap));
+}
+
+// A thread-safe LRU cache to store instances of BigintEnumType.
```

Comment:
Would you update documentation for this type to explain the usage of this cache?